### PR TITLE
Drive

### PR DESCRIPTION
**Pull request by RahulBansal1112**

### CODE REVIEW

**Comment by tsoim** [comment]
try rebasing on main first then pushing / creating the pull request? seems like you're trying to bring in changes that have already been merged

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -23,17 +23,18 @@
...
-	public static final double TANK_DRIVE_STRAIGHT_DRIVE_POWER_DIFF = 0.05;
+	public static final double TANK_DRIVE_STRAIGHT_DRIVE_POWER_DIFF = 0.1;
+	public static final double TRACKWIDTH_IN = 20.72;
 
 	//Path constants
-	public static final double RUN_4_LEAVE_TARMAC_DIST = 55.44;
-	public static final double RUN_4_BACK_TO_TARMAC_DIST = -83.577;
-	public static final double RUN_4_TURN_TO_HUB_ANGLE = 69.125;
-	public static final double RUN_4_BACK_TO_HUB_DIST = -23.128;
+	public static final double RUN_4_LEAVE_TARMAC_DIST = 55.699;
```
Comment (AG6GR) [inline src/main/java/frc/robot/Constants.java]: What are the units for these values?

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 	// Systems
 	private DriveFSMSystem driveFsmSystem;
-	private BallHandlingFSM ballSystem;
+	// private BallHandlingFSM ballSystem;
```
Comment (AG6GR) [inline src/main/java/frc/robot/Robot.java]: Let's leave all systems enabled on main branch, just comment for local testing.

File: src/main/java/frc/robot/trajectory/Kinematics.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+		double robotNewXPos = circleX + radius * Math.cos(Math.toRadians(beta));
+		double robotNewYPos = circleY + radius * Math.sin(Math.toRadians(beta));
+		
+		return new Point(robotNewXPos, robotNewYPos);
+	}
+
+	public static Point inversekinematics(double gyroHeading, Point robotPos, Point targetPos) {
+		Point newtargetPos = new Point(targetPos.getX() - robotPos.getX(), targetPos.getY() - robotPos.getY());
+
+		double mRC = Math.tan(Math.toRadians(gyroHeading));
```
Comment (AG6GR) [inline src/main/java/frc/robot/trajectory/Kinematics.java:74]: Can you rename these single letter variables, or add comments on what these represent? Also the units.

File: src/main/java/frc/robot/trajectory/Point.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package frc.robot.trajectory;
+
+public class Point {
```
Comment (AG6GR) [inline src/main/java/frc/robot/trajectory/Point.java:3]: Consider using the Translation2D class provided by WPILib https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/math/geometry/Translation2d.html

File: src/main/java/frc/robot/trajectory/Vector.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package frc.robot.trajectory;
+
+public class Vector {
```
Comment (AG6GR) [inline src/main/java/frc/robot/trajectory/Vector.java:3]: Same as with Point, could probably be replaced with Translation2D or Pose. https://docs.wpilib.org/en/stable/docs/software/advanced-controls/geometry/pose.html

File: src/main/java/frc/robot/trajectory/PurePursuit.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		double robotHeading = fsmSystem.getHeading();
+
+		// prevent OOB errors
+		int lookaheadPointIndex = lastClosestPointIndex + lookaheadDistance;
+		Point lookaheadPoint = lookaheadPointIndex < pathPoints.size()
+			? pathPoints.get(lookaheadPointIndex) : pathPoints.get(pathPoints.size() - 1);
+		return lookaheadPoint;
+	}
+
+	// private double calculateCurvature() {
```
Comment (AG6GR) [inline src/main/java/frc/robot/trajectory/PurePursuit.java]: Clean comments plz :)

File: src/main/java/frc/robot/trajectory/PurePursuit.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	private ArrayList<Point> pathPoints = new ArrayList<>();
+	private int lastClosestPointIndex = 0;
+	// lookahead point is this many points ahead of the closest point
+	private final int lookaheadDistance = 6;
+
+	// in inches
+	private static final double SPACING = 6.0;
+
+
+	public PurePursuit(ArrayList<Point> initialPoints, DriveFSMSystem fsmSystem) {
```
Comment (AG6GR) [inline src/main/java/frc/robot/trajectory/PurePursuit.java]: I'm not a fan of passing in fsmSystem to the constructor. DriveFSMSystem owns the PurePursuit instance, not the other way around. As far as I can tell, we only use fsmSystem to query values like getRobotPosArc and getheading. These should be passed in the method parameters when DriveFSMSystem calls the method.

**Review by AG6GR — CHANGES_REQUESTED** [review]

File: .wpilib/wpilib_preferences.json [hunk 8]
```diff
@@ -2,5 +2,5 @@
     "enableCppIntellisense": false,
     "currentLanguage": "java",
     "projectYear": "2022",
-    "teamNumber": 2473
+    "teamNumber": 2475
```
Comment (tsoim) [inline .wpilib/wpilib_preferences.json]: ?

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 9]
```diff
@@ -256,7 +308,9 @@ private void handleForwardOrBackwardState(TeleopInput input,
 		System.out.println("Inches: " + inches);
 		System.out.println("Error: " + error);
 
-		// Try This Next Time:
+		// Checks if either the encoder value is equal to the inches required (reached destination)
+		// or checks if the robot has hit a wall (encoder value is not chaning but motor power is
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java:316]: this case is not reflected in code? also typo for "chaning" -> "changing"

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 10]
```diff
@@ -256,7 +308,9 @@ private void handleForwardOrBackwardState(TeleopInput input,
...
+	// not zero)
 		if ((inches > 0 && error < Constants.ERR_THRESHOLD_STRAIGHT_IN)
 			|| (inches < 0 && error > -Constants.ERR_THRESHOLD_STRAIGHT_IN)) {
 			System.out.println("im here");
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java]: debug print statements shouldn't be included in PR, or at least make them more descriptive

File: src/main/java/frc/robot/trajectory/PurePursuit.java (hunk 6 above)
Comment (tsoim) [reply src/main/java/frc/robot/trajectory/PurePursuit.java]: consider git stash if there are things you want to save but not commit to main

https://www.atlassian.com/git/tutorials/saving-changes/git-stash

File: src/main/java/frc/robot/systems/DriveFSMSystem.java [hunk 11]
```diff
@@ -375,46 +432,109 @@ private void handleTeleOpState(TeleopInput input) {
...
-
-		prevGyroAngle = adjustedAngle;
-		prevEncoderPosArc = currentEncoderPos;
-		System.out.println("arc odo: (" + robotXPosArc + ", " + robotYPosArc + ")");
+		System.out.println("Target point: " + target.getX() + " " + target.getY());
+		Point motorSpeeds = Kinematics.inversekinematics(gyroAngle, robotPosArc, target, false);
+		frontLeftMotor.set(-motorSpeeds.getX() / 5);
+		frontRightMotor.set(motorSpeeds.getY() / 5);
+		backLeftMotor.set(-motorSpeeds.getX() / 5);
+		backRightMotor.set(motorSpeeds.getY() / 5);
```
Comment (tsoim) [inline src/main/java/frc/robot/systems/DriveFSMSystem.java]: magic num `5`?

File: .wpilib/wpilib_preferences.json (hunk 8 above)
Comment (RahulBansal1112) [reply .wpilib/wpilib_preferences.json]: We changed the old robot to have team number 2475 (so the ip for the router is different). So to test code on that robot, we needed to change the team number to 2475 (prob shouldn't have pushed that change to Drive though).

File: src/main/java/frc/robot/systems/DriveFSMSystem.java (hunk 11 above)
Comment (RahulBansal1112) [reply src/main/java/frc/robot/systems/DriveFSMSystem.java]: We will be cleaning up the code soon (we've been procrastinating on fixing the current checkstyle errors).

File: src/main/java/frc/robot/systems/DriveFSMSystem.java (hunk 10 above)
Comment (RahulBansal1112) [reply src/main/java/frc/robot/systems/DriveFSMSystem.java]: Yea we'll remove them

File: src/main/java/frc/robot/systems/DriveFSMSystem.java (hunk 11 above)
Comment (tsoim) [reply src/main/java/frc/robot/systems/DriveFSMSystem.java]: as a mentor I'm obligated to say that you should follow these best practices as you're coding, not after the fact :P please try to clean the code up before putting up a PR next time, it'll help reviewers focus on the actual code and not be confused/distracted by misc stuff :)

File: src/main/java/frc/robot/trajectory/Kinematics.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+		//center of the circle that defines the arc
+		Point center = new Point(0, 0);
+
+		//if mRC - mRT is close to 0, center.getX() will have a divide by 0 error
+		if (Math.abs(mRC - mRT) < Constants.ZERO_THRESHOLD) {
+			center.setX((newtargetPos.getX() * mRC + newtargetPos.getY() * mRC * mRT)
+				/ (2.0 * (mRC - mRT < 0 ? -1 : 1) * Constants.ZERO_THRESHOLD));
+			//if the heading is one way and the target point is 180 deg
+			//in the other direction, set the center of the circle at ~(0, 0)
+			if (gyroHeading > 270 && gyroHeading < 90 & newtargetPos.getX() < 0) {
```
Comment (AG6GR) [inline src/main/java/frc/robot/trajectory/Kinematics.java:109]: gyroHeading > 270 && gyroHeading < 90 is always false?

**Review by AG6GR — APPROVED** [review]

**Pull request merged**